File: BACK/src/Modules/Encounters/Explorer.Encounters.Core/Domain/Encounter.cs
using Explorer.BuildingBlocks.Core.Domain;
using Explorer.Encounters.Core.Domain.Enums;

namespace Explorer.Encounters.Core.Domain
{
    public class Encounter : Entity
    {
        public int UserId { get; init; }
        public String Name { get; init; }
        public String Description {  get; init; }
        public double Latitude {  get; private set; }
        public double Longitude { get; private set; }
        public int Xp {  get; init; }
        public EncounterStatus Status { get; init; }
        public EncounterType Type { get; init; }
        public double Range { get; init; }
        public string? Image {  get; init; }
        public double? ImageLatitude { get; private set; }
        public double? ImageLongitude { get; private set; }
        public int? PeopleCount { get; init; }
        public EncounterApprovalStatus ApprovalStatus { get; private set; }

        public Encounter() { }


        public Encounter(int userId, string name, string description, double latitude, double longitude, int xp, EncounterStatus status, EncounterType type, double range, string image, double imageLatitude, double imageLongitude, int peopleCount, EncounterApprovalStatus approvalStatus)
        {
            UserId = userId;
            Name = name;
            Description = description;
            Latitude = latitude;
            Longitude = longitude;
            Xp = xp;
            Status = status;
            Type = type;
            Range = range;
            Image = image;
            ImageLatitude = imageLatitude;
            ImageLongitude = imageLongitude;
            PeopleCount = peopleCount;
            ApprovalStatus = approvalStatus;
        }
        public void UpdateLocation(double latitude, double longitude)
        {
            Latitude = latitude;
            Longitude = longitude;
        }

        public void UpdateApprovalStatus(EncounterApprovalStatus approvalStatus) 
        {
            ApprovalStatus = approvalStatus;
        }

    }
}

File: ENCOUNTERS-MS/model/encounter.go
package model

import "github.com/google/uuid"

type EncounterStatus string

const (
	StatusActive   EncounterStatus = "ACTIVE"
	StatusDraft    EncounterStatus = "DRAFT"
	StatusArchived EncounterStatus = "ARCHIVED"
)

type EncounterApprovalStatus string

const (
	ApprovalPending        EncounterApprovalStatus = "PENDING"
	ApprovalSystemApproved EncounterApprovalStatus = "SYSTEM_APPROVED"
	ApprovalAdminApproved  EncounterApprovalStatus = "ADMIN_APPROVED"
	ApprovalDeclined       EncounterApprovalStatus = "DECLINED"
)

type EncounterType string

const (
	TypeSocial   EncounterType = "SOCIAL"
	TypeLocation EncounterType = "LOCATION"
	TypeMisc     EncounterType = "MISC"
)

type Encounter struct {
	ID             uuid.UUID
	UserID         int
	Name           string
	Description    string
	Latitude       float64
	Longitude      float64
	Xp             int
	Status         EncounterStatus
	Type           EncounterType
	Range          float64
	Image          string
	ImageLatitude  *float64
	ImageLongitude *float64
	PeopleCount    int
	ApprovalStatus EncounterApprovalStatus
}
